framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,152 @@
+     * return their receiver. This method will directly update the estimate of {@code a.b().c()} to
+     * include "foo". In addition, the estimates for the expressions {@code a.b()} and {@code a}
+     * would have their estimates updated to include "foo", because c and b (respectively) return
+     * their receivers. Note that due to what kind of values can be held in the store, this
+     * information is lost outside the method chain. That is, the returns-receiver propagated
+     * information is lost outside the expression in which the returns-receiver method invocations
+     * are nested.
+     *
+     * <p>As a concrete example, consider the Called Methods accumulation checker: if {@code build}
+     * requires a, b, and c to be called, then {@code foo.a().b().c().build();} will typecheck (they
+     * are in one fluent method chain), but {@code foo.a().b().c(); foo.build();} will not -- the
+     * store does not keep the information that a, b, and c have been called outside the chain.
+     * {@code foo}'s type will be {@code CalledMethods("a")}, because only {@code a()} was called
+     * directly on {@code foo}. For such code to typecheck, the Called Methods accumulation checker
+     * uses an additional rule: the return type of a receiver-returning method {@code rr()} is
+     * {@code CalledMethods("rr")}. This rule is implemented directly in the {@link
+     * org.checkerframework.framework.type.treeannotator.TreeAnnotator} subclass defined in the
+     * Called Methods type factory.
+     *
+     * @param node the node whose estimate should be expanded
+     * @param result the transfer result containing the store to be modified
+     * @param values the new accumulation values
+     */
+    public void accumulate(Node node, TransferResult<CFValue, CFStore> result, String... values) {
+        Tree tree = node.getTree();
+        if (tree == null) {
+            return;
+        }
+        AnnotatedTypeMirror oldType = typeFactory.getAnnotatedType(tree);
+        AnnotationMirror newAnno = getUnionAnno(oldType, Arrays.asList(values));
+        insertIntoStores(result, node, newAnno);
+
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
+            MethodInvocationNode methodInvocationNode = (MethodInvocationNode) node;
+            while (methodInvocationNode != null) {
+
+                Node receiver = methodInvocationNode.getTarget().getReceiver();
+
+                if (receiver == null || !typeFactory.returnsThis((MethodInvocationTree) tree)) {
+                    break;
+                }
+
+                // Note that this call doesn't do anything if receiver is a method call
+                // that is not deterministic, though the code below can still continue to recurse.
+                // (The code above tested that `tree` returns `this` but did not test `receiver`.)
+                insertIntoStores(result, receiver, newAnno);
```

Comment:
This use of `newAnno` looks wrong to me.  `newAnno` was computed earlier for a later method call, and if the recursion goes past a non-`@This`-returning call, then the value should be reset rather than carried across that call.  I added a failing test case, `m8`, that illustrates incorrect behavior.
Should the recursion go through non-`@This` returning methods at all?

It is also a bit suspicious that no call of `getUnionAnno` occurs.  Could you write a comment saying why it is not needed, and it's OK to do complete replacement rather than GLB?